liberum_core/src/swarm_runner/behaviour/kademlia.rs

```diff
@@ -1,87 +1,353 @@
+                    node = self.node.name,
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+            // Finish the query to prevent from triggering FinishedWithNoAdditionalRecord
+            self.swarm
+                .behaviour_mut()
+                .kademlia
+                .query_mut(&id)
+                .unwrap()
+                .finish();
+        }
+    }
+
+    fn handle_outbound_query_progressed_get_providers_no_additional_record(
+        &mut self,
+        id: QueryId,
+        _closest_peers: Vec<PeerId>,
+        _stats: QueryStats,
+        _step: ProgressStep,
+    ) {
+        debug!(
+            node = self.node.name,
+            "Get providers didn't find any new records"
+        );
+        if let Some(sender) = self.behaviour.pending_get_providers.remove(&id) {
+            let _ = sender.send(HashSet::new()).inspect_err(|e| {
+                debug!(
+                    qid = format!("{id}"),
+                    err = format!("{e:?}"),
+                    "Channel closed"
+                )
+            });
+        }
+    }
+}
+
+/// Methods to handle Kademlia InboundRequest events
+impl SwarmContext {
+    fn handle_inbound_request_put_record(
+        &mut self,
+        _source: PeerId,
+        _connection: ConnectionId,
+        record: Option<Record>,
+    ) {
+        debug!(node = self.node.name, "Kad Received PutRecord");
+        if record.is_none() {
+            warn!("Received PutRecord with no record");
+            return;
+        }
+        let record = record.unwrap();
+        let id = record.key.clone();
+
+        // save record to filem should use a VAULT here instead
+        self.put_record_into_vault(record);
+
+        // Start a query to be providing the file ID in kademlia
+        let qid = self
+            .swarm
+            .behaviour_mut()
+            .kademlia
+            .start_providing(id.clone());
+        if qid.is_err() {
+            debug!(
+                node = self.node.name,
+                err = format!("{qid:?}"),
+                "Failed to start providing file"
+            );
+            return;
+        }
+        let qid = qid.unwrap();
+
+        // We can't await for a response here because it would block the event loop
+        let (response_sender, _) = oneshot::channel();
+        self.behaviour
+            .pending_start_providing
+            .insert(qid, response_sender);
+    }
+
+    fn handle_inbound_request_add_provider(&mut self, record: Option<ProviderRecord>) {
+        match record {
+            Some(record) => {
+                self.swarm
+                    .behaviour_mut()
+                    .kademlia
+                    .store_mut()
+                    .add_provider(record.clone())
+                    .ok(); // TODO What if the providers amount is exceeded? How to ensure only the closest one are kept?
```

Comment:
I believe that GC was a solution to this problem, so cleaning up all the provider records every 24 hours and requiring providers to resend info about provided files across the network.

Reply:
There is a hard cap on how many providers the DHT will store (n=K=20 by default). This handler always adds a provider to the DHT when someone informs that they are a provider. I am not sure what happens when the cap is exceeded. Are the oldest ones dropped or what? Probably it's not that crucial...

Reply:
Are providers records treated like every other record?

Reply:
Provider records and the actual data records are two seperate things in kad. They just decided to use the size of the K-bucket as the maximum amount of providers of a given ID to be remembered.